keps/sig-network/20190415-Autopath API for clusterDNS.md

```diff
@@ -0,0 +1,104 @@
+Even if one of the several queries fails due, the DNS lookup on the client side will fail after a 5s timeout.
+
+### Goals
+Provide a solution to minimize number of DNS queries on the client side, for a DNS lookup. 
+This solution will preserve the current behavior that allows shortname lookups of Kubernetes resources. Make this solution configurable via the API.
+
+### Non-Goals
+Modifying the current Kubernetes DNS Schema is not a goal of this KEP.
+
+
+## Proposal
+
+This proposal introduces a new dnsPolicy "clusterFirstWithAutopath". This can be set by including the line below in podSpec:
+
+ ` dnsPolicy: clusterFirstWithAutopath`
+
+Using this mode will set the searchpath on the pod as `search.$NS.$SUFFIX.k`, where $NS is the namespace of the pod and $SUFFIX is the cluster suffix. k is a one-letter suffix to identify the domain name to be a query that needs search expansion. Since there are [no single-letter TLD so far](http://data.iana.org/TLD/tlds-alpha-by-domain.txt), this suffix will help identify queries needing search expansion. 
+
+This approach minimizes the number of DNS queries at client side to atmost 2(A, AAAA). The searchpath expansion logic moves to the server side. The server(clusterDNS - CoreDNS by default) will require additional metadata in order to complete searchpath expansion.
+
+The metadata can be attached as an EDNS0 option. We can define a new EDNS0 option - SearchPaths, option code: 15 that is currently [unassigned.](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-11) We would need to reserve this or use an option number from the experimental range - 65001 to 65534.
+The value of this option will be a comma-separated string consisting of all the searchpaths which are to be appended to the main query and looked up. This option can be useful outside of Kubernetes as well.
+
+Instead of modifying all client pod images to insert an EDNS0 option in their requests, we will put this logic in [Nodelocal DNSCache](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0030-nodelocal-dns-cache.md), which is a daemonset running a per-node DNS Cache.
+All pods using clusterDNS will send requests to the local instance running on the same node. This instance will look for the custom suffix that we defined above, in incoming query names. In case of a match, it will extract the namespace and cluster suffix values and construct the EDNS0 option with the searchpath strings.
+
+An optimization to consider: We can use the EDNS0 option to include a version number and metadata and move the logic of determining searchpaths to the server side. The benefit of this approach is that the size overhead of the EDNS0 option is minimal. But, it requires the server to have logic to determine searchpaths in addition to performing the expanded queries.
```

Comment:
If we want this to be widely adopted, this seems like a way to slow that down.  I don't really want to convene a committee to govern search path schema..

Reply:
I agree.

Reply:
Not sure I understand the concern. We are already defining an EDNS0 option that is just a list of search paths. My understanding of this request would be to instead populate that EDNS0 option with general metadata, and then it's up to the server to do with that what it will. I guess that is a more complex structure for the options, but either way we are defining our own option.

This is what I meant above when I said "implement more complex server-side DNS policy". For example, if the metadata included the source Pod namespace, and of course we know the destination namespace, we can enforce that these match (DNS-level multi-tenancy).

With this approach, I would consider this a generalization of the search path problem. "Search Path" is just another piece of meta-data.

Reply:
Each DNS vendor will need to interpret the metadata option to generate the searchpaths right? Explicitly listing it in the client makes it more straighforward for the server - Perform the queries with each of the paths listed in the option. That seems easier to standardize than "generate searchpaths given all this metadata"

Reply:
Yes, agreed. These things are not mutually exclusive in any case - we could have a ESN0 option for search path and another for general meta-data.

Reply:
good point